### PR TITLE
mruby-compiler: rebase the source array once an element index fills its operand

### PR DESCRIPTION
**Pull request by takumin**
## Summary

A multiple assignment and an array pattern read their elements with an index that is the third operand of a `BBB` instruction, which `OP_EXT1` to `OP_EXT3` do not widen, so past 256 elements the index wrapped and the wrong element was assigned or compared. The source array is now rebased on the elements not read yet every 255 of them, which keeps the index and the pre count of `OP_APOST` in range for any number of targets.

## Changes

| File                                   | What                                                                                                                                                                |
| -------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `mrbgems/mruby-compiler/src/codegen.c` | `gen_aref_scratch()` and `gen_aref_rebase()`, used by `gen_massignment()` and by both branches of the array pattern; a refusal for more than 255 post-splat targets |
| `test/t/codegen.rb`                    | 257 targets, 257 targets with a rest and post targets, an array pattern of 257 elements through both branches                                                       |
| `mrbgems/mruby-eval/test/eval.rb`      | 255 post-splat targets compile, 256 are refused                                                                                                                     |

### Rebasing the array instead of widening the operand

`OP_APOST(scratch, 255, 0)` drops the 255 elements already read and leaves the rest of the array in a register. The groups that follow are anchored on the elements that remain, not on the ones removed, so a rest and its post targets still choose the same elements after a rebase as before it, and the index starts over at zero. `gen_aref_scratch()` reserves the register only when there are more than 255 elements, so a shorter list generates exactly the instructions it generated before: disassembling the 110 Ruby sources of the core library gives 36,319 lines identical to master.

The pre count `OP_APOST` carries is the second operand, which `OP_EXT2` does widen, but it wrapped too, at 65,536 targets. It now counts only the targets since the last rebase and stays below 256, so it no longer needs `OP_EXT2` either.

### More than 255 targets after a rest

The post count cannot be rebased away. `OP_APOST` fills the post targets from the front of the array when the array is shorter than the targets and from the back when it is longer, and it decides which at run time from the length; each rebase would decide that over again for the group it split off, and the leftover of a short array would land in the wrong group. More than 255 targets after a rest is refused with `too many post-splat assignment targets` rather than assigned wrongly. Choosing between the two at run time instead would need the length of the array, which no instruction carries, and a count wider than an operand would have to be a new instruction, since widening the one `OP_APOST` has changes how the instructions already compiled decode. An array pattern is not refused for the same shape: it checks the length before it reads the elements, so its post elements are read with `OP_GETIDX` and an `mrb_int` index counted from the end.

CRuby compiles that, so this is a difference from it, on code mruby cannot run correctly today. Nothing else reaches it: as local variables the targets stop at 252 on `too many local variables`, and of the 5,226 multiple assignments in ruby/ruby, in every gem installed here and in this tree, none has more than 2 targets after a rest.

## Behaviour

```ruby
a = (0...300).to_a
@x0, @x1, @x2, ..., @x299 = a
@x299                             # CRuby: 299, mruby before: 43, mruby now: 299
```

```ruby
a = (0...300).to_a
case a
in [0, 1, 2, ..., 299] then :match
else :nomatch
end                               # CRuby: :match, mruby before: :nomatch, mruby now: :match
```

```ruby
a = (0...65537).to_a
@x, @x, ..., @x, *r = a           # 65,536 targets before the rest
r.size                            # CRuby: 1, mruby before: 65537, mruby now: 1
```

A rest followed by more than 255 targets was accepted and gave the rest the whole array and the post targets whatever the registers held; it is now refused at compile time.

## Size

`.text` of `bin/mruby`, `CC=gcc CXX=g++ LD=gcc`, gcc 13.3.0.

| Build            |    master |   this PR | delta |
| ---------------- | --------: | --------: | ----: |
| full-debug (-O0) | 2,003,862 | 2,004,470 |  +608 |
| bintest          | 1,401,206 | 1,401,942 |  +736 |
| cxx_abi          | 1,435,673 | 1,436,457 |  +784 |
| byte-string      | 1,362,582 | 1,363,318 |  +736 |
| ascii-ctype      | 1,385,606 | 1,386,342 |  +736 |
| no-float         | 1,328,750 | 1,329,550 |  +800 |
| no-bigint        | 1,286,950 | 1,287,686 |  +736 |

All of it is `codegen.o` (+736 in the library, +864 in the copy `mrbc` builds for itself): the two helpers are inlined into their callers, which grow by 323 bytes in `gen_massignment` and 464 in `codegen_pattern`. No other object changed, and the VM is untouched.

## Testing

| Build       | Total |    OK | Skip |  KO | Crash | Warning |
| ----------- | ----: | ----: | ---: | --: | ----: | ------: |
| host        | 2,828 | 2,754 |   74 |   0 |     0 |       0 |
| full-debug  | 3,076 | 3,065 |   11 |   0 |     0 |       0 |
| bintest     | 3,076 | 3,056 |   20 |   0 |     0 |       0 |
| cxx_abi     | 3,076 | 3,056 |   20 |   0 |     0 |       0 |
| byte-string | 2,988 | 2,914 |   74 |   0 |     0 |       0 |
| ascii-ctype | 3,060 | 3,038 |   22 |   0 |     0 |       0 |
| no-float    | 2,809 | 2,712 |   97 |   0 |     0 |       0 |
| no-bigint   | 3,025 | 2,992 |   33 |   0 |     0 |       0 |

`rake test:run:bin:bintest` is 148 with 147 OK, 1 skipped and no failure.

The new tests pin the element a target and a pattern element read past the operand: 257 targets, 257 targets followed by a rest and two post targets read from an array longer than the targets and from a two element one, an array pattern of 257 elements through the branch that reads the subject with `deconstruct` and through the branch that knows the subject's length, and the boundary of the refusal at 255 and 256 post-splat targets.

Beside the suite, the boundary was compared against CRuby 4.0.6 on 34 programs: multiple assignment at 254 to 258, 300, 511 to 513 and 65,536 targets, with a rest, with a rest and post targets, from an array shorter than the targets, from a value that is not an array, into a nested target list, into index targets and read as a value; and an array pattern at 256, 257 and 300 elements through both branches, matching, failing on the last element, and followed by a rest and a post element. Every one agrees except the refused shape.

## Environment

<details>

|          |                                                                   |
| -------- | ----------------------------------------------------------------- |
| OS       | Ubuntu 24.04.4 LTS                                                |
| Kernel   | Linux 7.0.0-31-generic x86_64                                     |
| CPU      | AMD Ryzen 9 5950X (16 cores, 32 threads)                          |
| Compiler | gcc 13.3.0 (Ubuntu 13.3.0-6ubuntu2~24.04.1)                       |
| binutils | GNU ld / ar 2.47.20260726                                         |
| CRuby    | ruby 4.0.6 (2026-07-14 revision 03b6d3f889) +PRISM [x86_64-linux] |

Compile lines, from `touch src/string.c; rake --verbose`, with `-MMD -c`, `-I` and `-o` dropped.

full-debug

```
gcc -g -O3 -Wall -Wundef -Werror-implicit-function-declaration -Wwrite-strings -g3 -O0 -DMRB_GC_STRESS -DMRB_USE_DEBUG_HOOK -DMRB_DEBUG -DMRB_REVISION_HEADER -DMRB_USE_BIGINT -DMRB_USE_COMPLEX -DHAVE_MRUBY_ENCODING_GEM -DMRB_UTF8_STRING -DHAVE_MRUBY_IO_GEM -DMRB_USE_RATIONAL -DHAVE_MRUBY_REGEXP_GEM -DMRB_USE_SET -DMRB_USE_TASK_SCHEDULER src/string.c
```

bintest

```
gcc -g -O3 -Wall -Wundef -Werror-implicit-function-declaration -Wwrite-strings -DMRB_GC_FIXED_ARENA -DMRB_REVISION_HEADER -DMRB_USE_BIGINT -DMRB_USE_COMPLEX -DHAVE_MRUBY_ENCODING_GEM -DMRB_UTF8_STRING -DHAVE_MRUBY_IO_GEM -DMRB_USE_RATIONAL -DHAVE_MRUBY_REGEXP_GEM -DMRB_USE_SET -DMRB_USE_TASK_SCHEDULER -DMRB_USE_DEBUG_HOOK src/string.c
```

cxx_abi (gcc compiles as C++, g++ only links)

```
gcc -O3 -Wall -Wundef -Wwrite-strings -x c++ -std=gnu++03 -DMRB_GC_FIXED_ARENA -DMRB_USE_CXX_EXCEPTION -DMRB_USE_CXX_ABI -DMRB_REVISION_HEADER -DMRB_USE_BIGINT -DMRB_USE_COMPLEX -DHAVE_MRUBY_ENCODING_GEM -DMRB_UTF8_STRING -DHAVE_MRUBY_IO_GEM -DMRB_USE_RATIONAL -DHAVE_MRUBY_REGEXP_GEM -DMRB_USE_SET -DMRB_USE_TASK_SCHEDULER src/string.c
```

byte-string

```
gcc -g -O3 -Wall -Wundef -Werror-implicit-function-declaration -Wwrite-strings -DMRB_REVISION_HEADER -DMRB_USE_BIGINT -DMRB_USE_COMPLEX -DHAVE_MRUBY_IO_GEM -DMRB_USE_RATIONAL -DHAVE_MRUBY_REGEXP_GEM -DMRB_USE_SET -DMRB_USE_TASK_SCHEDULER src/string.c
```

ascii-ctype

```
gcc -g -O3 -Wall -Wundef -Werror-implicit-function-declaration -Wwrite-strings -DMRB_USE_ASCII_CTYPE -DMRB_PROCESS_HAVE_GETRUSAGE=0 -DMRB_REVISION_HEADER -DMRB_USE_BIGINT -DMRB_USE_COMPLEX -DHAVE_MRUBY_ENCODING_GEM -DMRB_UTF8_STRING -DHAVE_MRUBY_IO_GEM -DMRB_USE_RATIONAL -DHAVE_MRUBY_REGEXP_GEM -DMRB_USE_SET -DMRB_USE_TASK_SCHEDULER src/string.c
```

no-float

```
gcc -g -O3 -Wall -Wundef -Werror-implicit-function-declaration -Wwrite-strings -DMRB_NO_FLOAT -DMRB_REVISION_HEADER -DMRB_USE_BIGINT -DHAVE_MRUBY_ENCODING_GEM -DMRB_UTF8_STRING -DHAVE_MRUBY_IO_GEM -DMRB_USE_RATIONAL -DHAVE_MRUBY_REGEXP_GEM -DMRB_USE_SET -DMRB_USE_TASK_SCHEDULER src/string.c
```

no-bigint

```
gcc -g -O3 -Wall -Wundef -Werror-implicit-function-declaration -Wwrite-strings -DMRB_REVISION_HEADER -DMRB_USE_COMPLEX -DHAVE_MRUBY_ENCODING_GEM -DMRB_UTF8_STRING -DHAVE_MRUBY_IO_GEM -DMRB_USE_RATIONAL -DHAVE_MRUBY_REGEXP_GEM -DMRB_USE_SET -DMRB_USE_TASK_SCHEDULER src/string.c
```

</details>

## Summary by CodeRabbit

* **Bug Fixes**
  * Multiple assignment and array pattern matching now work correctly with more than 255 elements before a splat.
  * Added a clear syntax error when more than 255 elements follow a splat, which exceeds the supported limit.
  * Improved handling across attribute targets, rest targets, insufficient values, method results, and array literals.

* **Tests**
  * Added coverage for large assignments and array patterns, including boundary cases at 255 and 256 elements.

### CODE REVIEW

File: mrbgems/mruby-compiler/src/codegen.c [hunk 1]
```diff
@@ -2728,30 +2728,75 @@ gen_call(mrc_codegen_scope *s, mrc_node *tree, int val, int safe, int recv_ready
   }
 }
 
+/* The index OP_AREF reads an element with is the third operand of a BBB
```
[the remark this answers is not in the document]
Comment (takumin) [reply mrbgems/mruby-compiler/src/codegen.c:2731]: The relative pronoun is left out, not the clause: it reads "the index [that] `OP_AREF` reads an element with", the same shape as "the array the index below counts from" further down this comment and "the register `gen_aref_rebase()` rebases into" in the next one. `with is` landing next to each other is a stumble to read, though, so I will take the rewording if a reviewer trips on it too.

File: mrbgems/mruby-compiler/src/codegen.c [hunk 2]
```diff
@@ -3139,15 +3191,26 @@ codegen_pattern_1(mrc_codegen_scope *s, mrc_node *pattern, int target, uint32_t
...
-          push();
-          /* Element is now at sp */
-          /* Match element pattern (elements are not known arrays) */
-          codegen_pattern(s, pat_arr->requireds.nodes[i], sp, fail_pos, -1, 0);
-          pop();
+        {
+          int base = arr_reg, idx = 0;
+          int scratch = gen_aref_scratch(s, pre_len);
+          for (i = 0; i < pre_len; i++) {
+            if (idx == 255) {
```
Comment (takumin) [inline mrbgems/mruby-compiler/src/codegen.c:3198]: Leaving the three walks as they are. Their bodies differ, `gen_assignment()` in `gen_massignment()` and `codegen_pattern()` with its own arguments in the two branches here, so one walker would have to take a callback or be a macro, and codegen.c has neither today. What repeats is the two lines that call `gen_aref_rebase()`; the rebase itself, and what it means for the groups that follow, is written once where that function is defined.
